components/script/script_thread.rs

```diff
@@ -3310,6 +3310,12 @@ impl ScriptThread {
         // TODO: This should only dirty nodes that are waiting for a web font to finish loading!
         document.dirty_all_nodes();
         document.window().add_pending_reflow();
+
+        // TODO: This is currently required because the handlers added to the promise exposed at
+        // `document.fonts.ready` are run by the event loop only when it performs a microtask
+        // checkpoint. Without the call below, this never happens and the promise is 'stuck'
+        // waiting to be resolved until another event forces a microtask checkpoint.
+        self.rendering_opportunity(pipeline_id);
```

Comment:
I don't think this is correct, because the mechanism that resolves the promise is not the microtask checkpoint(which by the way in our wrong implementation is called once after each iteration of the event-loop, so it is called after handling this message as well), but rather the call to `reflow`, which does happens in the "update the rendering"(not always), but I think in your case the reflow should happen after this(since a pending reflow is added above) as part of https://github.com/servo/servo/blob/66edef806579fd0b386f4ceba473e6a9f7d0ca34/components/script/script_thread.rs#L2011. Does the promise not resolve without noting a rendering opp here?

Also, see the other comment at `reflow`.

Reply:
> . Does the promise not resolve without noting a rendering opp here?

Yes, that is what we observed. This example exhibits the issue if `self.rendering_opportunity()` is removed.

```html
<!DOCTYPE html>
<html>
  <head>
  <script>
    document.fonts.ready.then(function() {
      console.log("Promise resolved with", arguments);
    })
  </script>
  <style>
    @font-face {
      font-family: 'FiraSans';
      src: url('https://google.com/'); /* Force font loading to fail due to sanitizer error */
      /* src: url('https://fonts.gstatic.com/s/firasans/v17/va9E4kDNxMZdWfMOD5Vvl4jL.woff2'); */
    }
    div {
      font-family: "FiraSans";
    }
  </style>
  </head>
  <body>
    <div>Test</div>
  </body>
</html>
```

However, we don't observe the issue when `src` is a valid font, so there maybe something more going on.

Reply:
Ok, thanks for clarifying. In that case, I think your fix is accidental, in that it works but doesn't address what I think is the underlying problem:

- The promise is resolved in one of the reflow calls at the end of the event-loop iteration: https://github.com/servo/servo/blob/8121c9883477cf4350e8dd8abdc3776548c11ea3/components/script/script_thread.rs#L1995
- The microtask checkpoint is done before, at https://github.com/servo/servo/blob/8121c9883477cf4350e8dd8abdc3776548c11ea3/components/script/script_thread.rs#L1974
- You are then queuing a task, accidentally the update the rendering one, but it could have been any task really,
- The task runs, followed by a micro tack checkpoint that runs the promise callback. 

So I think the more direct way of addressing the problem, is to resolve the promise right here. Even though we are not in a "task", the micro task checkpoint will run after this. (remaining question: do you need to reflow before the promise callback runs?)

Reply:
Forget about the above, we don't need the microtask checkpoint. It may still be an idea to resolve the promise already here(and see if the problem is not in the logic around it, like the check for all web fonts having been loaded already). 

Also a problem is that it is not clear that the document is fully loaded at this point, but that is a separate problem.